### PR TITLE
Revert custom errors with new `CustomRevert` library

### PR DESCRIPTION
**Pull request by shuhuiluo**
## Related Issue
Which issue does this pull request resolve?

Bytecode size.

## Description of changes
Solidity creates a new [`JUMPDEST`](https://www.evm.codes/#5b?fork=cancun) at a section of bytecode containing the revert logic every time `revert CustomError()` is used. This pull request introduces a `CustomRevert` library with a `revertWith` function as a unified `JUMPDEST` for handling the revert of custom errors without an argument hence reducing contract bytecode. It also demonstrates a syntax sugar for passing custom error selector to the function with `using CustomRevert for bytes4;`.

By replacing `revert CustomError()` with `CustomError.selector.revertWith()`, contract size is reduced by 9 bytes for each revert of custom errors without an argument, allowing for more aggressive optimizer runs.

### CODE REVIEW

File: src/NoDelegateCall.sol [hunk 1]
```diff
@@ -1,9 +1,13 @@
 // SPDX-License-Identifier: BUSL-1.1
 pragma solidity ^0.8.20;
 
+import {CustomRevert} from "./libraries/CustomRevert.sol";
```
Comment (hensha256) [inline src/NoDelegateCall.sol:4]: could do `using CustomRevert for bytes4 global`?
I actually think we should do that with more of our libs too

File: src/libraries/CustomRevert.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /// @dev Reverts with a custom error with two int24 arguments
+    /// @dev The free memory pointer is tampered with but it is fine since the call context is exited immediately
+    function revertWith(bytes4 selector, int24 value1, int24 value2) internal pure {
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0, selector)
+            mstore(0x04, value1)
+            mstore(0x24, value2)
```
Comment (hensha256) [inline src/libraries/CustomRevert.sol:51]: dont these top 4 bytes of the fmp need to be set back to 0?

Comment (hensha256) [reply src/libraries/CustomRevert.sol:51]: oh its reverting 💀

File: src/NoDelegateCall.sol (hunk 1 above)
Comment (shuhuiluo) [reply src/NoDelegateCall.sol:4]: Unfortunately the `global` keyword only works for user-defined types.
https://docs.soliditylang.org/en/latest/contracts.html#using-for

**Review by hensha256 — APPROVED** [review]

**Review by hensha256 — APPROVED** [review]

**Pull request merged**